src/modules/track_model/src/track_model.cc

```diff
@@ -1 +1,298 @@
+
+        // Check if the current block has a station and update deboarding
+        types::BlockId station_check = occupied_train_blocks_[i][0];
+
+        if (blocks_[station_check].has_station == 1)
+        {
+            uint16_t traindeb = trains_[i]->GetPassengersDeboarding();
+            SetPassengersDeboarding(i, traindeb);
+        }
+    }
+}
+
+
+
+types::Error SoftwareTrackModel::SetSwitchState(const types::BlockId block, const bool Switched)
+{
+    // Logic to set the Switch state for the specified block
+    if (blocks_.size() < block)
+    {
+        return types::ERROR_INVALID_BLOCK;
+    }
+    if (block <= 0)
+    {
+        return types::ERROR_INVALID_BLOCK;
+    }
+
+    blocks_[block].switched = Switched;
+
+    //update graph
+    if (Switched == 1)
+    {
+        //remove edge (previous connection)
+        graph.RemoveEdge(block, block + 1);
+
+        //add new edge (new connection)
+        types::BlockId newcon = blocks_[block].switch_connection;
+        graph.AddEdge(block, newcon, blocks_[newcon].length);
+    }
+    else
+    {
+        //remove edge (previous connection)
+        types::BlockId oldcon = blocks_[block].switch_connection;
+        graph.RemoveEdge(block, oldcon);
+
+        //add new edge (new connection)
+        graph.AddEdge(block, block + 1, blocks_[block + 1].length);
+    }
+
+    return types::ERROR_NONE;
+}
+
+types::Error SoftwareTrackModel::SetCrossingState(const types::BlockId block, const bool closed)
+{
+    return types::Error{ };
+}
+
+types::Error SoftwareTrackModel::SetRedTrafficLight(const types::BlockId block, const bool on)
+{
+    return types::ERROR_NONE;
+}
+
+types::Error SoftwareTrackModel::SetYellowTrafficLight(const types::BlockId block, const bool on)
+{
+    return types::ERROR_NONE;
+}
+
+types::Error SoftwareTrackModel::SetGreenTrafficLight(const types::BlockId block, const bool on)
+{
+    return types::ERROR_NONE;
+}
+
+types::Error SoftwareTrackModel::SetCommandedSpeed(const types::BlockId block, const types::MetersPerSecond speed)
+{
+    if (blocks_.size() < block)
+    {
+        return types::ERROR_INVALID_BLOCK;
+    }
+    if (block <= 0)
+    {
+        return types::ERROR_INVALID_BLOCK;
+    }
+
+    //loop through trains
+    for (int i = 0; i < trains_.size(); i++)
+    {
+        //does speed get sent to the block the (front of the) train is on?
+        if (occupied_train_blocks_[i][0] == block)
+        {
+            trains_[i]->SetCommandedSpeed(speed);
+
+            return types::ERROR_NONE;
+        }
+
+    }
+
+    return types::ERROR_INVALID_BLOCK;
+}
+
+types::Error SoftwareTrackModel::SetAuthority(const types::BlockId block, const types::Blocks authority)
+{
+    // Logic to set the authority
+
+    if (blocks_.size() < block)
+    {
+        return types::ERROR_INVALID_BLOCK;
+    }
+    if (block <= 0)
+    {
+        return types::ERROR_INVALID_BLOCK;
+    }
+
+    //loop through trains
+    for (int i = 0; i < trains_.size(); i++)
```

Comment:
Could be worth it to write a lookup function to get the train corresponding to a block instead of writing a loop every time